### DISCUSSION TITLE
How does Filament utilize the sh.txt file and under what circumstances?

### DISCUSSION THREAD

**Discussion by EnviralDesign** [google/filament#6625] [Q&A] [answered]
Greetings!
Been building, testing, and learning from the sample c++ projects, and so far have observed this behavior with hellopbr and gltf_viewer. I am not using the filament libraries and am reverse engineering and porting the shaders to another ecosystem (for context)

In the default build, there exists an IBL asset here:
`filament-1.31.6\out\build\x64-Debug\samples\assets\ibl\lightroom_14b`
Inside that directory is a skybox and ibl ktx file, as well as a sh.txt file.

Up until now I thought this was always used, however after inspecting the samples with renderdoc, I noticed that the iblSH frameUniform looks like this in all samples:
![image](https://user-images.githubusercontent.com/10091486/223828833-92fd4da2-7da0-4c0d-a9d4-9e06084de28e.png)

The only piece of information I could easily discern from the source code was from [IndirectLight.cpp](https://github.com/google/filament/blob/main/filament/src/details/IndirectLight.cpp):
```
struct IndirectLight::BuilderDetails {
    Texture const* mReflectionsMap = nullptr;
    Texture const* mIrradianceMap = nullptr;
    float3 mIrradianceCoefs[9] = { 65504.0f }; // magic value (max fp16) to indicate sh are not set
    mat3f mRotation = {};
    float mIntensity = FIndirectLight::DEFAULT_INTENSITY;
};
```

This would lead me to expect all of the sh values to be 65504.0 however it's just the ones in the first position of the array.

All this aside the results I'm seeing look correct so far as I can tell, comparing the output of a sample and my custom implementation, but why? Why does the sh.txt exist? Is it a quality thing, or a performance thing that can be enabled/disabled used or not?

Thanks!

**Comment by romainguy** [comment, marked as answer]
That file sh.txt is not used when using the KTX files. KTX files embed that information already.

**Reply by EnviralDesign** [reply]
how is that information passed into the shader though if not through `iblSH[9]` ?
I am looking at the raw inputs to the fragment shader in renderdoc, and those values are not set.

In my own implementation, I am trying to mirror the results I see in the sample, and I am not using ktx, but rather loading in the faces of the cubemap at different mips and reading it via a custom sampler function.

The results seem to be comparable, so unclear why

**Reply by romainguy** [reply]
Those values can be set optionally in the fragment. If they are set, we use them, otherwise we use one of the mip levels from the specular chain as the irradiance. This works better with very high dynamic range/high frequency environments.

**Reply by EnviralDesign** [reply]
Ok, thank you for the explanation, is one or the other better quality / performance aside from the high frequency environments? 
Or is this just a workflow / preference consideration other than that.
